curvefs/src/metaserver/copyset/copyset_node_manager.cpp

```diff
@@ -0,0 +1,231 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Date: Fri Aug  6 17:10:54 CST 2021
+ * Author: wuhanqing
+ */
+
+#include "curvefs/src/metaserver/copyset/copyset_node_manager.h"
+
+#include <brpc/server.h>
+
+#include <string>
+#include <utility>
+
+#include "absl/memory/memory.h"
+#include "curvefs/src/metaserver/copyset/copyset_reloader.h"
+#include "curvefs/src/metaserver/copyset/raft_cli_service2.h"
+#include "curvefs/src/metaserver/copyset/utils.h"
+#include "src/common/timeutility.h"
+
+namespace curvefs {
+namespace metaserver {
+namespace copyset {
+
+using ::curve::common::TimeUtility;
+
+CopysetNodeManager::CopysetNodeManager()
+    : options_(),
+      running_(false),
+      loadFinished_(false),
+      lock_(),
+      copysets_() {}
+
+bool CopysetNodeManager::IsLoadFinished() const {
+    return loadFinished_.load(std::memory_order_acquire);
+}
+
+bool CopysetNodeManager::DeleteCopysetNodeInternal(PoolId poolId,
+                                                   CopysetId copysetId,
+                                                   bool removeData) {
+    GroupId groupId = ToGroupId(poolId, copysetId);
+
+    // stop copyset node first
+    {
+        ReadLockGuard lock(lock_);
+        auto it = copysets_.find(groupId);
+        if (it != copysets_.end()) {
+            it->second->Stop();
+        } else {
+            LOG(WARNING) << "Delete copyset failed, copyset "
+                         << ToGroupIdString(poolId, copysetId) << " not found";
+            return false;
+        }
+    }
+
+    // remove copyset node
+    {
+        WriteLockGuard lock(lock_);
+        auto it = copysets_.find(groupId);
+        if (it != copysets_.end()) {
+            bool ret = true;
+            if (removeData) {
+                std::string copysetDataDir = it->second->GetCopysetDataDir();
+                if (!options_.trash->RecycleCopyset(copysetDataDir)) {
+                    LOG(WARNING) << "Recycle copyset remote data failed, "
+                                    "copyset data path: '"
+                                 << copysetDataDir << "'";
+                    ret = false;
+                }
+            }
+
+            copysets_.erase(it);
+            LOG(INFO) << "Delete copyset " << ToGroupIdString(poolId, copysetId)
+                      << " success";
+            return ret;
+        }
+    }
+
+    return false;
```

Comment:
why return false here?

Reply:
because the find within writelock may failed.